### DISCUSSION TITLE
What is the best way to model integer floor division?

### DISCUSSION THREAD

**Discussion by florianschanda** [cvc5/cvc5#9865] [Q&A]
I am attempting to generate VCs for a language where integer division semantics are python-like, i.e
| a | b | smtlib (div a b) | python a // b|
|--------|--------|--------|--------|
| 5 | 2 | 2 | 2 |
| -5 | 2 | -3 | -3 |
| 5 | -2 | -2 | -3 |
| -5 | -2 | 3 | 2 |

I am considering to implement this using the following function definition:
```lisp
(define-fun floordiv ((lhs Int) (rhs Int)) Int
  (ite (< rhs 0)
       (div (- lhs) (- rhs))
       (div lhs rhs)))
```

It gives the correct result, but is there a better way of doing this?

**Comment by ajreynol** [comment]
We don't have any builtin support for other kinds of division, so this seems reasonable to me. I suppose you could also try:
```
(define-fun floordiv ((lhs Int) (rhs Int)) Int
  (div 
    (ite (< rhs 0) (- lhs) lhs) 
    (ite (< rhs 0) (- rhs) rhs)))
```
But I don't have a clear idea which is better on first glance.

**Reply by florianschanda** [reply]
I suppose that one minimises the number of div operations as opposed to ite. I was wondering if you do any rewriting or simplification, then keeping the operands obvious would make it easier to do so, but I guess not?
